storage-proofs/src/zigzag/proof.rs

```diff
@@ -0,0 +1,727 @@
+            let pow = next_pow2(leafs);
+            let mut leaves_store = MerkleStore::new(pow);
+            populate_leaves::<_, <H as Hasher>::Function, _, std::iter::Map<_, _>>(
+                &mut leaves_store,
+                (0..leafs).map(|i| get_node::<H>(tree_data, i).unwrap()),
+            );
+
+            graph.merkle_tree_from_leaves(leaves_store, leafs)
+        };
+
+        #[allow(clippy::type_complexity)]
+        let (tree_d, tree_r_last, tree_c, comm_r, encodings, es, os): (
+            Tree<H>,
+            Tree<H>,
+            Tree<H>,
+            H::Domain,
+            Encodings<_>,
+            Vec<H::Domain>,
+            Vec<H::Domain>,
+        ) = crossbeam::thread::scope(|s| -> Result<_> {
+            // 1. Build the MerkleTree over the original data
+            trace!("build merkle tree for the original data");
+            let tree_d_handle = s.spawn(|_| match data_tree {
+                Some(t) => Ok(t),
+                None => build_tree(&data),
+            });
+            // 2. Encode all layers
+            trace!("encode layers");
+            let mut encodings: Vec<Vec<u8>> = Vec::with_capacity(layers - 1);
+            let mut current_graph = graph.clone();
+            let mut to_encode = data.to_vec();
+
+            for layer in 0..layers {
+                trace!("encoding (layer: {})", layer);
+                vde::encode(&current_graph, replica_id, &mut to_encode)?;
+                current_graph = Self::transform(&current_graph);
+
+                assert_eq!(to_encode.len(), NODE_SIZE * nodes_count);
+
+                if layer != layers - 1 {
+                    let p = to_encode.clone();
+                    encodings.push(p);
+                }
+            }
+
+            assert_eq!(encodings.len(), layers - 1);
+
+            let encodings = Encodings::<H>::new(encodings);
+
+            let r_last = to_encode;
+
+            // Construct final replica commitment
+            let tree_r_last_handle = s.spawn(move |_| build_tree(&r_last));
+
+            // 3. Construct Column Commitments
+            let odd_columns = (0..nodes_count)
+                .into_par_iter()
+                .map(|x| encodings.odd_column(x));
+
+            let even_columns = (0..nodes_count)
+                .into_par_iter()
+                .map(|x| encodings.even_column(graph.inv_index(x)));
+
+            // O_i = H( e_i^(1) || .. )
+            let os = odd_columns
+                .map(|c| c.map(|c| Fr::from(c.hash()).into()))
+                .collect::<Result<Vec<H::Domain>>>()?;
+
+            // E_i = H( e_\bar{i}^(2) || .. )
+            let es = even_columns
+                .map(|c| c.map(|c| Fr::from(c.hash()).into()))
+                .collect::<Result<Vec<H::Domain>>>()?;
+
+            // C_i = H(O_i || E_i)
+            let cs = os
+                .par_iter()
+                .zip(es.par_iter())
+                .flat_map(|(o_i, e_i)| hash2(o_i, e_i).as_ref().to_vec())
```

Comment:
Probably not following the new ZigZag paper correctly, but isn't the even column index inverted for `C_i`?